Improvements to the Roku Media Player

reedriley opened the pull request:
## Description:
My Roku wasn't being detected correctly as described in #5697.  Rather than just downgrading the script, I selectively reverted part to get it working.  Also, I tweaked the detection logic for "idle" (screensaver) states to work correctly with non-default screensavers.

**Related issue (if applicable):** fixes #5697

## Checklist:

If the code communicates with devices, web services, or third-party tools:
  - [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

homeassistant commented:
Hello @reedriley,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`3d6070c3b73386934a9bba6612afa8270427aa4f`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;6265&#x2F;commits&#x2F;3d6070c3b73386934a9bba6612afa8270427aa4f) - This commit has something that looks like an email address (reedriley@nox.local). Maybe try linking that to GitHub?.
- [`030d50060b34f6a0c22f8499529227e2fef1e7a5`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;6265&#x2F;commits&#x2F;030d50060b34f6a0c22f8499529227e2fef1e7a5) - This commit has something that looks like an email address (reedriley@nox.local). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

armills commented:
So the play_pause problem was introduced during the async media_player conversion, similar to #5282. I'll get a fix in for this.

Why are you rolling back other PRs? e.g. #5475

reedriley commented:
I have three issues with my Roku:
 1) Didn't autodetect properly out of the box until I reverted a small portion of the media player script for Roku to an earlier version,
 2) Didn't detect custom screensavers as "Idle" properly,
 3) Didn't implement pause/play correctly.

I can check tonight to see whether the latest trunk works properly (issue #1), and rebase/clean up my work on top of it if appropriate.  That part of the fix is the only one where I was intentionally rolling back changes.

armills commented:
OK, thanks for the clarification:
1. Instead of reverting #5475, have the name default to userdevicename, then fall back to the old serial number logic if userdevicename is not defined.
2. Currently, we're using [`self.current_app.is_screensaver`](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/media_player/roku.py#L135) to detect screensavers. If it's not working correctly can you please submit the fix upstream?
3. This should be fixed with #6291. This was actually a regression from the async conversion. Can you test this on your system?

reedriley committed:
Add fallback for name if userdevicename isn't set using old serialnumber logic

reedriley commented:
1. Good idea, I've switched to that.
2. `self.current_app.is_screensaver` works beautifully.  That wasn't present in the version I was working off of, so I've reverted that part of my change.
3. #6291 also works great, I've reverted that part of my changes as well.

Thanks @armills!

armills commented:
Great! The change looks good, but this branch got a lot of extra commits pulled into it. The easiest way to fix is copying out your `roku.py` start a new branch based on the latest dev, paste in your file, then do a git diff to make sure only your intended changes are going in. Then do a `git push -f  origin/roku`, assuming origin is the name of your remote.

reedriley commented:
@armills: I'm not sure why github thinks there are 21 commits in this pull request.  I've already rebased everything on top of the tag `0.39.1` and done a `--force` push to clean it up.  Let me rebase off of `dev` to see if that looks cleaner.

For my own curiosity, do you have a wiki page or something describing your branching model?

armills commented:
It would be a good idea to have a documentation for the branching. It's been slowly evolving to where we are now. Basically what happens right now is a few days before the release a release branch is cut from dev, so in this example `release-0.39`. New commits continue to get put into dev, and bug-fixes are cherry-picked onto the release branch.

So since you rebased on the release branch but are merging into dev you've got the extra commits. Rebasing on dev will fix it.

emlove reviewed (approved):
LGTM. Going to page @robbiet480 since he did #5475
